File: GuildedRose/src/com/gildedrose/items/ItemDecorator.java
package com.gildedrose.items;

public abstract class ItemDecorator {

	// Constants
	protected final int MIN_QUALITY = 0;
	protected final int MAX_QUALITY = 50;
	
	// Members
	private Item itemToDecorate;
	
	// Constructors
	protected ItemDecorator(Item itemToDecorate)
	{
		this.itemToDecorate = itemToDecorate;
	}
	
	// Properties
	public String getName()
	{
		return this.itemToDecorate.name;
	}
	
	public void setName(String name)
	{
		this.itemToDecorate.name = name;
	}
	
	public int getSellIn()
	{
		return this.itemToDecorate.sellIn;
	}
	
	public void setSellIn(int sellIn)
	{
		this.itemToDecorate.sellIn = sellIn;
	}
	
	public int getQuality()
	{
		return this.itemToDecorate.quality;
	}
	
	public void setQuality(int quality)
	{
		this.itemToDecorate.quality = quality;
	}

	// Read-only properties
    protected boolean IsSellInEnded()
    {
        return this.itemToDecorate.sellIn <= 0;
    }

    protected boolean IsSellInEvenNumber()
    {
        return this.itemToDecorate.sellIn % 2 == 0;
    }

    protected boolean IsSellInOddNumber()
    {
        return !this.IsSellInEvenNumber();
    }
	
    // Public methods
	public void update()
	{
		this.updateQuality();
		this.updateSellIn();
	}
	
	// Protected methods
	protected abstract void updateQuality();
	
	protected void updateSellIn()
	{
		this.itemToDecorate.sellIn--;
	}

	protected void decreaseQualityOnce()
	{
		this.itemToDecorate.quality -= 1;
		if (this.itemToDecorate.quality < MIN_QUALITY) this.itemToDecorate.quality = MIN_QUALITY;
	}
	
	protected void decreaseQualityTwice()
	{
		this.itemToDecorate.quality -= 2;
		if (this.itemToDecorate.quality < MIN_QUALITY) this.itemToDecorate.quality = MIN_QUALITY;
	}
	
	protected void decreaseQualityQuarce()
	{
		this.itemToDecorate.quality -= 4;
		if (this.itemToDecorate.quality < MIN_QUALITY) this.itemToDecorate.quality = MIN_QUALITY;
	}
	
	protected void increaseQualityOnce()
	{
		this.itemToDecorate.quality += 1;
		if (this.itemToDecorate.quality > MAX_QUALITY) this.itemToDecorate.quality = MAX_QUALITY;
	}
	
	protected void increaseQualityTwice()
	{
		this.itemToDecorate.quality += 2;
		if (this.itemToDecorate.quality > MAX_QUALITY) this.itemToDecorate.quality = MAX_QUALITY;
	}
	
	protected void increaseQualityThrice()
	{
		this.itemToDecorate.quality += 3;
		if (this.itemToDecorate.quality > MAX_QUALITY) this.itemToDecorate.quality = MAX_QUALITY;
	}

}

File: GildedRose.Net/GildedRose.Net/Items/ItemDecorator.cs
namespace GildedRose.Net.Items
{
    public abstract class ItemDecorator
    {

        // Constants
        protected const int MIN_QUALITY = 0;
        protected const int MAX_QUALITY = 50;
    
        // Members
        private Item itemToDecorate;

        // Constructors
        protected ItemDecorator(Item itemToDecorate)
        {
            this.itemToDecorate = itemToDecorate;
        }
        
        // Properties
        public string Name
        {
            get
            {
                return this.itemToDecorate.Name;
            }
            set
            {
                this.itemToDecorate.Name = value;
            }
        }
        
        public int SellIn
        {
            get
            {
                return this.itemToDecorate.SellIn;
            }
            set
            {
                this.itemToDecorate.SellIn = value;
            }
        }
        
        public int Quality
        {
            get
            {
                return this.itemToDecorate.Quality;
            }
            set
            {
                this.itemToDecorate.Quality = value;
            }
        }
    
        // Read-only properties
        protected bool SellInEnded
        {
            get {
                return this.SellIn <= 0;
            }
        }
        
        protected bool SellInEvenNumber
        {
            get {
                return this.SellIn % 2 == 0;
            }
        }
        
        protected bool SellInOddNumber
        {
            get {
                return this.SellIn % 2 != 0;
            }
        }
        
        // Public methods
        public void Update()
        {
            this.UpdateQuality();
            this.UpdateSellIn();
        }
        
        // Protected methods
        protected abstract void UpdateQuality();
        
        protected virtual void UpdateSellIn()
        {
            this.SellIn--;
        }
        
        protected void DecreaseQualityOnce()
        {
            this.Quality -= 1;
            if (this.Quality < MIN_QUALITY) this.Quality = MIN_QUALITY;
        }
        
        protected void DecreaseQualityTwice()
        {
            this.Quality -= 2;
            if (this.Quality < MIN_QUALITY) this.Quality = MIN_QUALITY;
        }
        
        protected void DecreaseQualityQuarce()
        {
            this.Quality -= 4;
            if (this.Quality < MIN_QUALITY) this.Quality = MIN_QUALITY;
        }
        
        protected void IncreaseQualityOnce()
        {
            this.Quality += 1;
            if (this.Quality > MAX_QUALITY) this.Quality = MAX_QUALITY;
        }
        
        protected void IncreaseQualityTwice()
        {
            this.Quality += 2;
            if (this.Quality > MAX_QUALITY) this.Quality = MAX_QUALITY;
        }
        
        protected void IncreaseQualityThrice()
        {
            this.Quality += 3;
            if (this.Quality > MAX_QUALITY) this.Quality = MAX_QUALITY;
        }

    }
}